swagger.yaml

```diff
@@ -3498,7 +3498,88 @@ paths:
           schema:
             $ref: '#/definitions/error'
 
+  /v2/clusters/{cluster_id}/ignored-validations:
+    get:
+      tags:
+        - installer
+      description: Fetch the validations which are to be ignored for this cluster.
+      operationId: v2GetIgnoredValidations
+      parameters:
+        - in: path
+          name: cluster_id
+          description: The cluster whose failing validations should be ignored according to this list.
+          type: string
+          format: uuid
+          required: true
+      responses:
+        "200":
+          description: Success.
+          schema:
+            $ref: '#/definitions/ignored-validations'
+        "400":
+          description: Error.
+          schema:
+            $ref: '#/definitions/error'
+        "404":
+          description: Error.
+          schema:
+            $ref: '#/definitions/error'
+    put:
+      tags:
+        - installer
+      description: Register the validations which are to be ignored for this cluster.
+      operationId: v2SetIgnoredValidations
+      parameters:
+        - in: path
+          name: cluster_id
+          description: The cluster whose failing validations should be ignored according to this list.
+          type: string
+          format: uuid
+          required: true
+        - in: body
+          name: ignored_validations
+          description: The validations to be ignored.
+          required: true
+          schema:
+            $ref: '#/definitions/ignored-validations'
+      responses:
+        "201":
+          description: Success.
+          schema:
+            $ref: '#/definitions/cluster' 
+        "400":
+          description: Error.
+          schema:
+            $ref: '#/definitions/error'
+        "404":
+          description: Error.
+          schema:
+            $ref: '#/definitions/error'
+        "401":
+          description: Unauthorized.
+          schema:
+            $ref: '#/definitions/infra_error'
+        "403":
+          description: Forbidden.
+          schema:
+            $ref: '#/definitions/infra_error'
+        "500":
+          description: Error.
+          schema:
+            $ref: '#/definitions/error'
+
 definitions:
+  ignored-validations:
```

Comment:
I don't like that the host and cluster ignored validations are bundled up like this, clusters are one entity, hosts are another entity, each of them should have its own field.

Mostly because hosts are not always bound to clusters and can move between them

Reply:
This is a cluster level setting and should not be applied on a "per host" basis, that's not the intention here.

The intention is that for very specific customers, we would like to enable them to mark specific validations as skippable across the board. What this setting means is "during the installation of this cluster, whenever you encounter this host validation, ignore it..."

If we go the other way, registering this on a per host basis, it's possible but it would complicate the usage of this feature for the customer. 

Open to change my mind if there is broader support for changing in this way.

Reply:
Hosts have validations even when detached from clusters, how would users skip those then?

Reply:
At least, they do as far as I know. I may be wrong about this

Reply:
Checking this point, I didn't make it clear in the enhancement doc but unbound clusters are a non-goal for this feature. This is really for those customers who have a "more or less" viable cluster ready to install but are blocked by one validation that would allow the installation to proceed if they could just skip it.

Reply:
I'm sorry I missed this earlier, but @omertuc has a good point.  We may want to expose this functionality in our Kubernetes API, and in late binding, we do have some host validations that run before binding.  Worse, we block binding if a host has failing validations.

Reply:
kube-api is one thing late binding is another thing, after all the controller will call what ever it needs. 
the downside of having it in the host level is that users will need to do it per host, on the other had having it in the infra-env level will mean that users that use late binding (we don't have many of those as far as i know) will affect all the hosts in the infra-env. 
So i assume that having the field for each host is more flexible. but it's sounds like an overkill for something that we don't want to officially support. 
for me for now it's enough to have it on the cluster level and if we see a demand we can decide on how to improve, it's not an api that we need to maintain and not break.

Reply:
> that users will need to do it per host

Not necessarily, we could have a hybrid solution where hosts inherit the validation-skips from their cluster, and also have their own list

Reply:
A hybrid solution can also be done in parts, i.e. first merge this PR as is, then add per-host skips in the future

Reply:
The downside of a hybrid solution is that it's much more complex in the backend - two tiers of validation skips that have to be merged

Reply:
The suggestion from @omertuc sounds reasonable as part of a strategy where we see whether or not there is demand to expand the functionality further.

Reply:
"we will need to think how to add it to kube-api without having that directly in the spec because we don't want to expose it as part of the official api" -> What about a custom CRD that accepts a cluster reference?

Reply:
i was thinking about annotations, but for now i would not even add it to kube-api at this point.
reference is still something that you will have in the spec.

Reply:
So for now, this REST API should do the job?

Reply:
So for now, this REST API should do the job?

Reply:
yes, lets start with that and see if it's used.

Reply:
> improve only on demand

It will be harder to change later once it's in the DB.  Why not fix it now?

Reply:
> > improve only on demand
> 
> It will be harder to change later once it's in the DB. Why not fix it now?

Assuming we're going for an eventual hybrid solution, you still need the clusters to specify both host and cluster validations, so clusters can affect all of their hosts in bulk, making it simpler for users

Reply:
> I think a hybrid solution would be even more complicated because we would need to deal with different values on the host vs on the cluster.

Yes, that's the point I was making [here](https://github.com/openshift/assisted-service/pull/4927#discussion_r1094429262)

Although the combination is a simple union(set(host_skips) | set(cluster_host_skips))

Reply:
@filanov looking for your further input here...

Reply:
if it will be on the host level we will need to handle cases where host is reused, isn't that an overkill? 
Lets remember the goal, it's to skip validation in a very specific cases to enable starting instlalation, and it's not supported api. 
late binding can be a non goal, if the infra-env have a cluster reference host will not be unbound so the validations are not relevant there. 

> It will be harder to change later once it's in the DB. Why not fix it now?
what is the concern about the DB? cluster will have a list regardless which solution we have.

Reply:
> what is the concern about the DB? cluster will have a list regardless which solution we have.
Either the cluster has two lists (one for cluster and one for host validations), o the cluster resource has one list and the host resource has one list.

What about a host with some hardware that we're not validating properly, or someone wants to install it with a smaller disk?  The validation should be ignored regardless of the cluster that the host is bound to.

Reply:
Maybe i'll ask it in a different way, with the existing solution besides late binding, do we have any case that will prevent the user to start installation if he is using the api that suggested here? and the second question is, what in this solution blocking us from making improvements later (Avishay said DB, not sure i understand why)?

Reply:
I think it will be relatively little effort to change now.  Later it will be more difficult.  We'll add it to the host, then people will ask why it's both on the host and on the cluster, then we'll remove it from the cluster.

Reply:
another option is to dispose of this whole host vs cluster flexibility, make it a single flag "ignore all validations (with the exception of the ones we can't ignore)"

So I could get rid of the JSON fields and replace with a boolean.

@ori-amizur made this suggestion, also sounds reasonable?

Reply:
> I think it will be relatively little effort to change now. Later it will be more difficult. We'll add it to the host, then people will ask why it's both on the host and on the cluster, then we'll remove it from the cluster.

OK lets go with that then. 
two different apis, one for host and one for cluster. in hat case we have it per host so users will have a bit more work but ti's fine because the api is not recommended anyway

Reply:
@filanov before we go ahead, should we consider the suggestion from @ori-amizur ?

Reply:
Then let's just stick with what we have in this PR and not overcomplicate things

Reply:
yes please

Reply:
Resolving this conversation. 
The current api will provide a good solution for all the users, we are ignoring late binding for this api. 
as described it's a patch solution for blocked users and not a supported api.